### ISSUE TITLE
"Transport not open" still an issue

### ISSUE THREAD

**Issue by cloudhack** [Quobject/EngineIoClientDotNet#63]
I keep having issues with my application crashing and exiting.
Using EngineIoClientDotNet version 1.0.6
.NET version 4.6.1
After adding 
`            AppDomain.CurrentDomain.UnhandledException += (sender, args) =>
            {
                Debug.WriteLine("********************************************************");
                Debug.WriteLine("Unhandled exception: " + args.ExceptionObject);
                Debug.WriteLine("********************************************************");
            };`

i get this stack trace
`********************************************************
Unhandled exception: Quobject.EngineIoClientDotNet.Client.EngineIOException: Transport not open
   at Quobject.EngineIoClientDotNet.Client.Transport.Send(ImmutableList`1 packets)
   at Quobject.EngineIoClientDotNet.Client.Socket.Flush()
   at Quobject.EngineIoClientDotNet.Client.Socket.SendPacket(Packet packet, Action fn)
   at Quobject.EngineIoClientDotNet.Client.Socket.<SetPing>b__65_0()
   at Quobject.EngineIoClientDotNet.Thread.EasyTimer.<>c__DisplayClass2_0.<SetTimeout>b__0()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.Tasks.AwaitTaskContinuation.RunCallback(ContextCallback callback, Object state, Task& currentTask)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
********************************************************
An unhandled exception of type 'Quobject.EngineIoClientDotNet.Client.EngineIOException' occurred in mscorlib.dll
Additional information: Transport not open`

I would expect this to be catched and the Socket.EVENT_DISCONNECT to be triggered instead ?

**Comment by LolliDepp** [comment]
I'm having the same issue, when it happens the client has usually been running for hours, if not for days.
One strange thing I did notice, is one time I got this error on multiple clients, on different remote locations, but all the clients crashed at about the same time.
Could this be an exception due to some server feature not properly handled?

I enabled logs on the EngineIoClientDotNet library, this is what I've found in XunitTrace.log near the time of the crash

The unhandled exception was caught at 10:29:29

> 2018-11-06 10:27:09 040 [] EngineIoClientDotNet-Socket.cs:SetPing#571 EasyTimer SetPing start
2018-11-06 10:27:09 040 [] EngineIoClientDotNet-Socket.cs:Flush#448 ReadyState=OPEN Transport.Writeable=True Upgrading=False WriteBuffer.Count=1
2018-11-06 10:27:09 040 [] EngineIoClientDotNet-Socket.cs:Flush#448 Flush 1 packets in socket
2018-11-06 10:27:09 040 [] EngineIoClientDotNet-Transport.cs:Send#153 Send called with packets.Count: 1
2018-11-06 10:27:09 040 [] EngineIoClientDotNet-Transport.cs:Send#153 Transport open, wiriting packet
2018-11-06 10:27:09 040 [] EngineIoClientDotNet-Transport.cs:Writable#41 Writable: False sid=nBaNFi1CTG6uTMI1AATD
2018-11-06 10:27:09 040 [] EngineIoClientDotNet-Transport.cs:Writable#41 Writable: True sid=nBaNFi1CTG6uTMI1AATD
2018-11-06 10:27:09 040 [] EngineIoClientDotNet-Socket.cs:SetPing#571 EasyTimer SetPing finish
2018-11-06 10:27:09 071 [] EngineIoClientDotNet-WebSocket.cs:ws_MessageReceived#104 ws_MessageReceived e.Message= 3
2018-11-06 10:27:09 071 [] EngineIoClientDotNet-Socket.cs:OnPacket#468 socket received: type 'pong', data ''
2018-11-06 10:27:09 071 [] EngineIoClientDotNet-Socket.cs:SetPing#566 writing ping packet - expecting pong within 5000ms
2018-11-06 10:27:34 078 [] EngineIoClientDotNet-Socket.cs:SetPing#571 EasyTimer SetPing start
2018-11-06 10:27:34 078 [] EngineIoClientDotNet-Socket.cs:Flush#448 ReadyState=OPEN Transport.Writeable=True Upgrading=False WriteBuffer.Count=1
2018-11-06 10:27:34 078 [] EngineIoClientDotNet-Socket.cs:Flush#448 Flush 1 packets in socket
2018-11-06 10:27:34 078 [] EngineIoClientDotNet-Transport.cs:Send#153 Send called with packets.Count: 1
2018-11-06 10:27:34 078 [] EngineIoClientDotNet-Transport.cs:Send#153 Transport open, wiriting packet
2018-11-06 10:27:34 078 [] EngineIoClientDotNet-Transport.cs:Writable#41 Writable: False sid=nBaNFi1CTG6uTMI1AATD
2018-11-06 10:27:34 078 [] EngineIoClientDotNet-Transport.cs:Writable#41 Writable: True sid=nBaNFi1CTG6uTMI1AATD
2018-11-06 10:27:34 078 [] EngineIoClientDotNet-Socket.cs:SetPing#571 EasyTimer SetPing finish
2018-11-06 10:27:34 109 [] EngineIoClientDotNet-WebSocket.cs:ws_MessageReceived#104 ws_MessageReceived e.Message= 3
2018-11-06 10:27:34 109 [] EngineIoClientDotNet-Socket.cs:OnPacket#468 socket received: type 'pong', data ''
2018-11-06 10:27:34 109 [] EngineIoClientDotNet-Socket.cs:SetPing#566 writing ping packet - expecting pong within 5000ms
2018-11-06 10:27:59 116 [] EngineIoClientDotNet-Socket.cs:SetPing#571 EasyTimer SetPing start
2018-11-06 10:27:59 116 [] EngineIoClientDotNet-Socket.cs:Flush#448 ReadyState=OPEN Transport.Writeable=True Upgrading=False WriteBuffer.Count=1
2018-11-06 10:27:59 116 [] EngineIoClientDotNet-Socket.cs:Flush#448 Flush 1 packets in socket
2018-11-06 10:27:59 116 [] EngineIoClientDotNet-Transport.cs:Send#153 Send called with packets.Count: 1
2018-11-06 10:27:59 116 [] EngineIoClientDotNet-Transport.cs:Send#153 Transport open, wiriting packet
2018-11-06 10:27:59 116 [] EngineIoClientDotNet-Transport.cs:Writable#41 Writable: False sid=nBaNFi1CTG6uTMI1AATD
2018-11-06 10:27:59 116 [] EngineIoClientDotNet-Transport.cs:Writable#41 Writable: True sid=nBaNFi1CTG6uTMI1AATD
2018-11-06 10:27:59 116 [] EngineIoClientDotNet-Socket.cs:SetPing#571 EasyTimer SetPing finish
2018-11-06 10:27:59 147 [] EngineIoClientDotNet-WebSocket.cs:ws_MessageReceived#104 ws_MessageReceived e.Message= 3
2018-11-06 10:27:59 147 [] EngineIoClientDotNet-Socket.cs:OnPacket#468 socket received: type 'pong', data ''
2018-11-06 10:27:59 147 [] EngineIoClientDotNet-Socket.cs:SetPing#566 writing ping packet - expecting pong within 5000ms
2018-11-06 10:28:24 154 [] EngineIoClientDotNet-Socket.cs:SetPing#571 EasyTimer SetPing start
2018-11-06 10:28:24 154 [] EngineIoClientDotNet-Socket.cs:Flush#448 ReadyState=OPEN Transport.Writeable=True Upgrading=False WriteBuffer.Count=1
2018-11-06 10:28:24 154 [] EngineIoClientDotNet-Socket.cs:Flush#448 Flush 1 packets in socket
2018-11-06 10:28:24 154 [] EngineIoClientDotNet-Transport.cs:Send#153 Send called with packets.Count: 1
2018-11-06 10:28:24 154 [] EngineIoClientDotNet-Transport.cs:Send#153 Transport open, wiriting packet
2018-11-06 10:28:24 154 [] EngineIoClientDotNet-Transport.cs:Writable#41 Writable: False sid=nBaNFi1CTG6uTMI1AATD
2018-11-06 10:28:24 154 [] EngineIoClientDotNet-Transport.cs:Writable#41 Writable: True sid=nBaNFi1CTG6uTMI1AATD
2018-11-06 10:28:24 154 [] EngineIoClientDotNet-Socket.cs:SetPing#571 EasyTimer SetPing finish
2018-11-06 10:28:24 185 [] EngineIoClientDotNet-WebSocket.cs:ws_MessageReceived#104 ws_MessageReceived e.Message= 3
2018-11-06 10:28:24 185 [] EngineIoClientDotNet-Socket.cs:OnPacket#468 socket received: type 'pong', data ''
2018-11-06 10:28:24 185 [] EngineIoClientDotNet-Socket.cs:SetPing#566 writing ping packet - expecting pong within 5000ms
2018-11-06 10:28:49 192 [] EngineIoClientDotNet-Socket.cs:SetPing#571 EasyTimer SetPing start
2018-11-06 10:28:49 332 [] EngineIoClientDotNet-Socket.cs:Flush#448 ReadyState=OPEN Transport.Writeable=True Upgrading=False WriteBuffer.Count=1
2018-11-06 10:28:49 348 [] EngineIoClientDotNet-Socket.cs:Flush#448 Flush 1 packets in socket
2018-11-06 10:28:49 348 [] EngineIoClientDotNet-Transport.cs:Send#153 Send called with packets.Count: 1
2018-11-06 10:28:49 348 [] EngineIoClientDotNet-Transport.cs:Send#153 Transport open, wiriting packet
2018-11-06 10:28:49 348 [] EngineIoClientDotNet-Transport.cs:Writable#41 Writable: False sid=nBaNFi1CTG6uTMI1AATD
2018-11-06 10:28:49 348 [] EngineIoClientDotNet-Transport.cs:Writable#41 Writable: True sid=nBaNFi1CTG6uTMI1AATD
2018-11-06 10:28:49 348 [] EngineIoClientDotNet-Socket.cs:SetPing#571 EasyTimer SetPing finish
2018-11-06 10:28:49 395 [] EngineIoClientDotNet-WebSocket.cs:ws_MessageReceived#104 ws_MessageReceived e.Message= 3
2018-11-06 10:28:49 395 [] EngineIoClientDotNet-Socket.cs:OnPacket#468 socket received: type 'pong', data ''
2018-11-06 10:28:49 395 [] EngineIoClientDotNet-Socket.cs:SetPing#566 writing ping packet - expecting pong within 5000ms
2018-11-06 10:29:24 120 [] EngineIoClientDotNet-Socket.cs:OnHeartbeat#1198 EasyTimer OnHeartbeat start
2018-11-06 10:29:24 120 [] EngineIoClientDotNet-Socket.cs:OnHeartbeat#1198 EasyTimer OnHeartbeat start
2018-11-06 10:29:24 120 [] EngineIoClientDotNet-Socket.cs:SetPing#571 EasyTimer SetPing start
2018-11-06 10:29:24 307 [] EngineIoClientDotNet-Socket.cs:OnClose#1118 OnClose socket close with reason: ping timeout
2018-11-06 10:29:24 495 [] EngineIoClientDotNet-WebSocket.cs:ws_Closed#92 ws_Closed
2018-11-06 10:29:24 557 [] EngineIoClientDotNet-Socket.cs:Flush#448 ReadyState=OPEN Transport.Writeable=True Upgrading=False WriteBuffer.Count=1
2018-11-06 10:29:24 588 [] EngineIoClientDotNet-Socket.cs:Flush#448 Flush 1 packets in socket
2018-11-06 10:29:24 588 [] EngineIoClientDotNet-Transport.cs:Send#153 Send called with packets.Count: 1
2018-11-06 10:29:24 651 [] EngineIoClientDotNet-Transport.cs:Send#153 Transport not open
2018-11-06 10:29:26 757 [] EngineIoClientDotNet-Socket.cs:OnHeartbeat#1198 EasyTimer OnHeartbeat finish
2018-11-06 10:29:27 973 [] EngineIoClientDotNet-Socket.cs:SetTransport#242 SetTransport setting transport 'polling'
2018-11-06 10:29:27 989 [] EngineIoClientDotNet-Socket.cs:Open#192 Task.Run Open start
2018-11-06 10:29:28 005 [] EngineIoClientDotNet-PollingXHR.cs:DoPoll#177 xhr DoPoll
2018-11-06 10:29:28 051 [] EngineIoClientDotNet-PollingXHR.cs:Create#269 xhr open GET: *redacted url*
2018-11-06 10:29:28 145 [] EngineIoClientDotNet-PollingXHR.cs:Create#315 Task.Run Create start
2018-11-06 10:29:29 034 [] EngineIoClientDotNet-PollingXHR.cs:Create#269 Xhr.GetResponse 
2018-11-06 10:29:29 237 [] EngineIoClientDotNet-PollingXHR.cs:OnData#390 OnData string = 96:0{"sid":"3VBi6a7LPlDZ5qu4AATN","upgrades":["websocket"],"pingInterval":25000,"pingTimeout":5000}2:40
2018-11-06 10:29:29 237 [] EngineIoClientDotNet-Polling.cs:_onData#198 polling got data 96:0{"sid":"3VBi6a7LPlDZ5qu4AATN","upgrades":["websocket"],"pingInterval":25000,"pingTimeout":5000}2:40
2018-11-06 10:29:29 253 [] EngineIoClientDotNet-Transport.cs:Writable#41 Writable: True sid=
2018-11-06 10:29:29 253 [] EngineIoClientDotNet-Socket.cs:OnPacket#468 socket received: type 'open', data '{"sid":"3VBi6a7LPlDZ5qu4AATN","upgrades":["websocket"],"pingInterval":25000,"pingTimeout":5000}'
2018-11-06 10:29:29 580 [] EngineIoClientDotNet-Socket.cs:OnHandshake#510 OnHandshake
2018-11-06 10:29:29 643 [] EngineIoClientDotNet-Socket.cs:Flush#448 ReadyState=OPEN Transport.Writeable=True Upgrading=False WriteBuffer.Count=0
2018-11-06 10:29:29 643 [] EngineIoClientDotNet-Socket.cs:Flush#448 Flush Not Send
2018-11-06 10:29:29 705 [] EngineIoClientDotNet-Socket.cs:Flush#448 ReadyState=OPEN Transport.Writeable=True Upgrading=False WriteBuffer.Count=1
2018-11-06 10:29:29 705 [] EngineIoClientDotNet-Socket.cs:Flush#448 Flush 1 packets in socket
2018-11-06 10:29:29 705 [] EngineIoClientDotNet-Transport.cs:Send#153 Send called with packets.Count: 1
2018-11-06 10:29:29 705 [] EngineIoClientDotNet-Transport.cs:Send#153 Transport open, wiriting packet
2018-11-06 10:29:29 705 [] EngineIoClientDotNet-Polling.cs:Write#307 Write packets.Count = 1
2018-11-06 10:29:29 721 [] EngineIoClientDotNet-Transport.cs:Writable#41 Writable: False sid=3VBi6a7LPlDZ5qu4AATN
2018-11-06 10:29:29 845 [] EngineIoClientDotNet-PollingXHR.cs:DoWrite#105 DoWrite data = System.Byte[]
2018-11-06 10:29:29 861 [] EngineIoClientDotNet-PollingXHR.cs:Create#269 xhr open POST: *redacted url*
2018-11-06 10:29:30 329 [] EngineIoClientDotNet-PollingXHR.cs:Create#315 Task.Run Create start
2018-11-06 10:29:30 485 [] EngineIoClientDotNet-PollingXHR.cs:Create#269 Xhr.GetResponse 
2018-11-06 10:29:30 485 [] EngineIoClientDotNet-PollingXHR.cs:OnData#390 OnData string = ok
2018-11-06 10:29:30 485 [] EngineIoClientDotNet-Transport.cs:Writable#41 Writable: True sid=3VBi6a7LPlDZ5qu4AATN
2018-11-06 10:29:30 641 [] EngineIoClientDotNet-PollingXHR.cs:Create#315 Task.Run Create finish
2018-11-06 10:29:30 641 [] EngineIoClientDotNet-Socket.cs:OnOpen#704 OnOpen starting upgrade probes
2018-11-06 10:29:30 657 [] EngineIoClientDotNet-Socket.cs:Probe#728 Probe probing transport 'websocket'
2018-11-06 10:29:30 703 [] EngineIoClientDotNet-WebSocket.cs:DoOpen#38 DoOpen uri =ws://*redacted url*
2018-11-06 10:29:30 813 [] EngineIoClientDotNet-Socket.cs:SetPing#566 writing ping packet - expecting pong within 5000ms
2018-11-06 10:29:30 828 [] EngineIoClientDotNet-Socket.cs:OnPacket#468 socket received: type 'message', data '0'
2018-11-06 10:29:31 078 [] EngineIoClientDotNet-Polling.cs:_onData#198 ReadyState != ReadyStateEnum.CLOSED
2018-11-06 10:29:31 078 [] EngineIoClientDotNet-PollingXHR.cs:DoPoll#177 xhr DoPoll
2018-11-06 10:29:31 171 [] EngineIoClientDotNet-PollingXHR.cs:Create#269 xhr open GET: *redacted url*
2018-11-06 10:29:31 171 [] EngineIoClientDotNet-PollingXHR.cs:Create#315 Task.Run Create start
2018-11-06 10:29:31 405 [] EngineIoClientDotNet-PollingXHR.cs:Create#269 Xhr.GetResponse 
2018-11-06 10:29:31 437 [] EngineIoClientDotNet-PollingXHR.cs:OnData#390 OnData string = 14:40/storetasks,
2018-11-06 10:29:31 437 [] EngineIoClientDotNet-Polling.cs:_onData#198 polling got data 14:40/storetasks,
2018-11-06 10:29:31 437 [] EngineIoClientDotNet-Socket.cs:OnPacket#468 socket received: type 'message', data '0/storetasks,'
2018-11-06 10:29:31 702 [] EngineIoClientDotNet-WebSocket.cs:ws_Opened#85 ws_Opened True
2018-11-06 10:29:31 702 [] EngineIoClientDotNet-Transport.cs:Writable#41 Writable: True sid=3VBi6a7LPlDZ5qu4AATN
2018-11-06 10:29:31 702 [] EngineIoClientDotNet-Transport.cs:Send#153 Send called with packets.Count: 1
2018-11-06 10:29:31 702 [] EngineIoClientDotNet-Transport.cs:Send#153 Transport open, wiriting packet
2018-11-06 10:29:31 702 [] EngineIoClientDotNet-Transport.cs:Writable#41 Writable: False sid=3VBi6a7LPlDZ5qu4AATN
2018-11-06 10:29:31 702 [] EngineIoClientDotNet-Transport.cs:Writable#41 Writable: True sid=3VBi6a7LPlDZ5qu4AATN
2018-11-06 10:29:31 733 [] EngineIoClientDotNet-WebSocket.cs:ws_MessageReceived#104 ws_MessageReceived e.Message= 3probe
2018-11-06 10:29:32 950 [] EngineIoClientDotNet-Socket.cs:WaitForUpgrade#675 Wait for upgrade timeout
2018-11-06 10:29:33 028 [] EngineIoClientDotNet-Socket.cs:Flush#448 ReadyState=OPEN Transport.Writeable=True Upgrading=True WriteBuffer.Count=1
2018-11-06 10:29:33 028 [] EngineIoClientDotNet-Socket.cs:Flush#448 Flush Not Send
2018-11-06 10:29:33 028 [] EngineIoClientDotNet-Polling.cs:_onData#198 ReadyState != ReadyStateEnum.CLOSED
2018-11-06 10:29:33 028 [] EngineIoClientDotNet-Socket.cs:Call#827 changing transport and sending upgrade packet
2018-11-06 10:29:33 043 [] EngineIoClientDotNet-Socket.cs:SetTransport#242 SetTransport setting transport 'websocket'
2018-11-06 10:29:33 043 [] EngineIoClientDotNet-Socket.cs:SetTransport#242 SetTransport clearing existing transport 'websocket'
2018-11-06 10:29:33 043 [] EngineIoClientDotNet-Transport.cs:Send#153 Send called with packets.Count: 1
2018-11-06 10:29:33 043 [] EngineIoClientDotNet-Transport.cs:Send#153 Transport open, wiriting packet
2018-11-06 10:29:33 043 [] EngineIoClientDotNet-Transport.cs:Writable#41 Writable: False sid=3VBi6a7LPlDZ5qu4AATN
2018-11-06 10:29:33 043 [] EngineIoClientDotNet-Transport.cs:Writable#41 Writable: True sid=3VBi6a7LPlDZ5qu4AATN
2018-11-06 10:29:33 043 [] EngineIoClientDotNet-Socket.cs:Flush#448 ReadyState=OPEN Transport.Writeable=True Upgrading=True WriteBuffer.Count=1
2018-11-06 10:29:33 043 [] EngineIoClientDotNet-Socket.cs:Flush#448 Flush Not Send
2018-11-06 10:29:33 043 [] EngineIoClientDotNet-Socket.cs:Flush#448 ReadyState=OPEN Transport.Writeable=True Upgrading=False WriteBuffer.Count=1
2018-11-06 10:29:33 043 [] EngineIoClientDotNet-Socket.cs:Flush#448 Flush 1 packets in socket
2018-11-06 10:29:33 043 [] EngineIoClientDotNet-Transport.cs:Send#153 Send called with packets.Count: 1
2018-11-06 10:29:33 043 [] EngineIoClientDotNet-Transport.cs:Send#153 Transport open, wiriting packet
2018-11-06 10:29:33 043 [] EngineIoClientDotNet-Transport.cs:Writable#41 Writable: False sid=3VBi6a7LPlDZ5qu4AATN
2018-11-06 10:29:33 043 [] EngineIoClientDotNet-Transport.cs:Writable#41 Writable: True sid=3VBi6a7LPlDZ5qu4AATN
2018-11-06 10:29:33 043 [] EngineIoClientDotNet-Polling.cs:_onData#198 ignoring poll - transport state PAUSED
2018-11-06 10:29:33 043 [] EngineIoClientDotNet-PollingXHR.cs:Create#315 Task.Run Create finish
2018-11-06 10:29:33 059 [] EngineIoClientDotNet-PollingXHR.cs:Create#315 Task.Run Create finish
2018-11-06 10:29:33 059 [] EngineIoClientDotNet-Socket.cs:Open#192 Task.Run Open finish
2018-11-06 10:29:33 402 [] EngineIoClientDotNet-WebSocket.cs:ws_MessageReceived#104 ws_MessageReceived e.Message= 42/storetasks,4219["assigntask",{ *redacted JSON here* }]
